src/EntryPoints/Specials/SpecialApproverCategories.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace ProfessionalWiki\PageApprovals\EntryPoints\Specials;
+
+use MediaWiki\MediaWikiServices;
+use ProfessionalWiki\PageApprovals\Adapters\DatabaseApproverRepository;
+use ProfessionalWiki\PageApprovals\Adapters\DatabaseApproversCategoriesLookup;
+use SpecialPage;
+use PermissionsError;
+use LightnCandy\LightnCandy;
+use WebRequest;
+
+class SpecialApproverCategories extends SpecialPage {
+
+	public function __construct() {
+		parent::__construct( 'ApproverCategories' );
+	}
+
+	public function isListed(): bool {
+		return $this->isAdmin(); // TODO: Add right permission checks
+	}
+
+	public function execute( $subPage ): void {
+		if ( !$this->isAdmin() ) {
+			throw new PermissionsError( 'sysop' );
+		}
+
+		$db = MediaWikiServices::getInstance()->getDBLoadBalancer()->getConnection( DB_PRIMARY );
+		$approversCategoriesLookup = new DatabaseApproversCategoriesLookup( $db );
+		$databaseApproverRepository = new DatabaseApproverRepository( $db );
+
+		$approversCategories = $approversCategoriesLookup->getApproversWithCategories();
+
+		$request = $this->getRequest();
+		if ( $request->wasPosted() ) {
+			$this->handlePostRequest( $request, $approversCategories, $databaseApproverRepository );
+			$this->getOutput()->redirect( $this->getPageTitle()->getLocalURL() );
+			return;
+		}
+
+		$this->renderHtml( $approversCategories );
+	}
+
+	private function isAdmin(): bool {
+		$userGroups = MediaWikiServices::getInstance()->getUserGroupManager()->getUserGroups( $this->getUser() );
+		return in_array( 'sysop', $userGroups );
+	}
+
+	/**
+	 * @param array<array{username: string, userId: int, categories: string[]}> $approversCategories
+	 */
+	private function handlePostRequest( WebRequest $request, array $approversCategories, DatabaseApproverRepository $databaseApproverRepository ): void {
+		$action = $request->getText( 'action' );
+		$username = $request->getText( 'username' );
+		$category = $request->getText( 'category' );
+
+		$userData = array_filter( $approversCategories, fn( $approver ) => $approver['username'] === $username );
+		if ( empty( $userData ) || empty( $category ) ) {
+			return;
+		}
+		$userData = array_values( $userData )[0];
+		$userId = $userData['userId'];
+		$currentCategories = $userData['categories'];
+
+		if ( $action === 'add' ) {
+			$currentCategories[] = $category;
+		} elseif ( $action === 'delete' ) {
+			$currentCategories = array_filter( $currentCategories, fn( string $cat ) => $cat !== $category );
+		}
+
+		$databaseApproverRepository->setApproverCategories( $userId, $currentCategories );
```

Comment:
This method should not know what implementation it uses.
`ApproverRepository $approverRepository`

Reply:
What's the best way to handle something like this? I am not sure what params would new method take if we create one.

Should it be like this?

`private function setApproverCategories( $userId, $newCategory, $categories, $action ) {
		if ( $action === 'add' ) {
			$categories[] = $newCategory;
		} elseif ( $action === 'delete' ) {
			$categories = array_filter( $categories, fn( string $cat ) => $cat !== $newCategory );
		} 
        $this->databaseApproverRepository->setApproverCategories( $userId, $categories );
	}
	`
or
	`$this->databaseApproverRepository->setApproverCategories( $userId, $currentCategories );`
	in the same method.